steps/expval.py

```diff
@@ -0,0 +1,159 @@
+from zquantum.core.measurement import expectation_values_to_real
+from zquantum.core.utils import create_object, save_list
+
+
+def run_circuit_and_get_expval(
+    backend_specs: str,
+    circuit: str,
+    operators: str,
+):
+    """Executes a circuit to obtain the expectation value of an operator on a
+    given backend.
+
+    All Orquestra backend interface calls used in this function are standard
+    methods of the ``QuantumBackend`` and ``QuantumSimulator`` interfaces as
+    defined in the ``z-quantum-core`` repository.
+
+    There are two computation modes: sampling and exact. Expectation values are
+    computed by post-processing samples when an Orquestra ``QuantumBackend`` is
+    used or the number of samples was specified for a ``QuantumSimulator``
+    backend. When the number of samples isn't specified, ``QuantumSimulator``
+    backends run in exact mode.
+
+    Args:
+        backend_specs (str): the Orquestra backend specification in a json
+            representation
+        circuit (str): the circuit represented as an OpenQASM 2.0 program
+        operators (str): the operator in an ``openfermion.QubitOperator``
+            or ``openfermion.IsingOperator`` representation
+    """
+    backend_specs = json.loads(backend_specs)
+    operators = json.loads(operators)
+
+    backend = create_object(backend_specs)
+
+    # 1. Parse circuit
+    qc = QuantumCircuit.from_qasm_str(circuit)
+
+    # 2. Create operators
+    ops = []
+    if backend.n_samples is not None:
+        # Operators for Backend/Simulator in sampling mode
+        for op in operators:
+            ops.append(IsingOperator(op))
+    else:
+        # Operators for Simulator exact mode
+        for op in operators:
+            ops.append(QubitOperator(op))
+
+    # 2.+1
+    # Activate the qubits that are measured but were not acted on
+    # By applying the identity
+    # Note: this is a temporary logic subject to be removed once supported by
+    # Orquestra
+
+    # Get the active qubits of the circuit
+    active_qubits = []
+    for instr in qc.data:
+        instruction_qubits = [qubit.index for qubit in instr[1]]
+        active_qubits.extend(instruction_qubits)
+
+    active_qubits = set(active_qubits)
+
+    # Get the qubits we'd like to measure
+    # Data for identities is not stored, need to account for empty terms
+    op_qubits = [term[0][0] for op in ops for term in op.terms if term]
+
+    need_to_activate = set(op_qubits) - active_qubits
+    if need_to_activate:
+        for qubit in need_to_activate:
+            # Apply the identity
+            qc.id(qubit)
+
+    # If there are still no instructions, apply identity to the first qubit
+    # Can happen for an empty circuit when measuring the identity operator
+    if not qc.data:
+        qc.id(qc.qubits[0])
+
+    # Convert to zquantum.core.circuit.Circuit
+    circuit = Circuit(qc)
+
+    # 3. Expval
+    results = _get_expval(backend, circuit, ops)
```

Comment:
I'm assuming the number of shots is passed implicitly as `backend.n_samples`?

Reply:
Yes, the number of samples is stored with
```python
    backend_specs = json.loads(backend_specs)
    backend = create_object(backend_specs)
```